File: frontend/lib/models/Course.dart
import 'dart:convert';

Course courseFromJson(String str) => Course.fromJson(json.decode(str));

String courseToJson(Course data) => json.encode(data.toJson());

class Course {
    Course({
        this.data,
    });

    List<CourseData> data;

    factory Course.fromJson(Map<String, dynamic> json) => Course(
        data: List<CourseData>.from(json["data"].map((x) => CourseData.fromJson(x))),
    );

    Map<String, dynamic> toJson() => {
        "data": List<dynamic>.from(data.map((x) => x.toJson())),
    };
}

class CourseData {
    CourseData({
        this.isHighestRated,
        this.isRecentlyAdded,
        this.isTrendingCourse,
        this.isAccepted,
        this.ratedBy,
        this.hasBeenCommented,
        this.comments,
        this.id,
        this.courseName,
        this.courseDescription,
        this.channelName,
        this.courseUrl,
        this.coursePic,
        this.category,
        this.courseRatings,
        this.coursePublishedOn,
        this.createdAt,
        this.updatedAt,
        this.v,
    });

    bool isHighestRated;
    bool isRecentlyAdded;
    bool isTrendingCourse;
    bool isAccepted;
    List<String> ratedBy;
    bool hasBeenCommented;
    List<Comment> comments;
    String id;
    String courseName;
    String courseDescription;
    String channelName;
    String courseUrl;
    String coursePic;
    String category;
    String courseRatings;
    String coursePublishedOn;
    DateTime createdAt;
    DateTime updatedAt;
    int v;

    factory CourseData.fromJson(Map<String, dynamic> json) => CourseData(
        isHighestRated: json["isHighestRated"],
        isRecentlyAdded: json["isRecentlyAdded"],
        isTrendingCourse: json["isTrendingCourse"],
        isAccepted: json["isAccepted"],
        ratedBy: List<String>.from(json["ratedBy"].map((x) => x)),
        hasBeenCommented: json["hasBeenCommented"],
        comments: List<Comment>.from(json["comments"].map((x) => Comment.fromJson(x))),
        id: json["_id"],
        courseName: json["courseName"],
        courseDescription: json["courseDescription"],
        channelName: json["channelName"],
        courseUrl: json["courseUrl"],
        coursePic: json["coursePic"],
        category: json["category"],
        courseRatings: json["courseRatings"],
        coursePublishedOn: json["coursePublishedOn"],
        createdAt: DateTime.parse(json["createdAt"]),
        updatedAt: DateTime.parse(json["updatedAt"]),
        v: json["__v"],
    );

    Map<String, dynamic> toJson() => {
        "isHighestRated": isHighestRated,
        "isRecentlyAdded": isRecentlyAdded,
        "isTrendingCourse": isTrendingCourse,
        "isAccepted": isAccepted,
        "ratedBy": List<dynamic>.from(ratedBy.map((x) => x)),
        "hasBeenCommented": hasBeenCommented,
        "comments": List<dynamic>.from(comments.map((x) => x.toJson())),
        "_id": id,
        "courseName": courseName,
        "courseDescription": courseDescription,
        "channelName": channelName,
        "courseUrl": courseUrl,
        "coursePic": coursePic,
        "category": category,
        "courseRatings": courseRatings,
        "coursePublishedOn": coursePublishedOn,
        "createdAt": createdAt.toIso8601String(),
        "updatedAt": updatedAt.toIso8601String(),
        "__v": v,
    };
}

class Comment {
  
    Comment({
        this.id,
        this.commentedText,
        this.commentedBy,
    });

    String id;
    String commentedText;
    String commentedBy;

    factory Comment.fromJson(Map<String, dynamic> json) => Comment(
        id: json["_id"],
        commentedText: json["commentedText"],
        commentedBy: json["commentedBy"],
    );

    Map<String, dynamic> toJson() => {
        "_id": id,
        "commentedText": commentedText,
        "commentedBy": commentedBy,
    };
}

File: server/routes/course.js
const express = require('express');
const router = express.Router();

const Courses = require("../models/course");
const Comment = require('../models/comment');
const Category = require('../models/category');
const User = require("../models/user");

const reqLogin = require('../middleware/reqLogin');
const reqAdmin = require('../middleware/reqAdmin');


router.get('/allcourses', (req, res) => {
    Courses.find({})
        .sort('-createdAt')
        .populate("comments","commentedText commentedBy")        
        .then(courses => {
            res.status(201).json({ data: courses });
        })
        .catch((err) => {
            console.log(err)
        })
});

router.get('/getTrending', async(req, res) => {
    await Courses.find({ isTrendingCourse: true })
        .sort('-createdAt')
        .populate("comments","commentedText commentedBy")
        .then((course) => {
            res.status(201).json({data:course});
        })
        .catch((err) => {
            console.log(err);
        })
});

router.get('/getHighest',async(req, res) => {
    await Courses.find({ isHighestRated: true })
        .sort('-createdAt')
        .populate("comments","commentedText commentedBy")
        .then(course => {
            res.status(201).json({data:course})
        })
        .catch((e) => {
            console.log(e);
        })
});


router.get('/getRecent', async(req, res) => {
    await Courses.find({ isRecentlyAdded: true })
        .sort('-createdAt')
        .populate("comments","commentedText commentedBy")
        .then(course => {
            res.status(201).json({data:course})
        })
        .catch((e) => {
            console.log(e);
        })
});



router.post('/addCourse', reqLogin,async(req, res) => {
    const course = new Courses(req.body);
    await course.save().then(result => {
        res.status(201).json(result);
    }).catch((e) => {
        console.log(e);
    });
});


router.post('/addReqCourse',reqLogin,async(req,res)=>{
    const {courseName,courseDescription,coursePic,channelName,courseUrl,courseRatings,category} = req.body;
    if(!courseName || !courseDescription || !coursePic || !channelName || !courseUrl){
        res.status(401).json({message:"Please Add All The Feilds"})
    }
    else{
    const isAccepted=false;
    const userCourse = new Courses({
        courseName,
        courseDescription,
        coursePic,
        channelName,
        courseUrl,
        courseRatings,
        isAccepted,
        category
    });
    await userCourse.save().then((s)=>{
        res.status(201).json(s)
    })
}
});




router.put('/ratecourse', reqLogin, async (req, res) => {
    const updateRatings = await Courses.findByIdAndUpdate(req.body.courseId, {
        courseRatings: req.body.newRatings,
        $push: {
            ratings: req.user._id
        }
    });
    const addInUser = await User.findByIdAndUpdate(req.user._id, {
        $push: {
            ratingsGiven: req.body.courseId
        }
    });


    if (updateRatings && addInUser) {
        res.status(201).json(updateRatings);
    }
});


router.post('/searchcourse', (req, res) => {
    let coursePattern = new RegExp("^" + req.body.query)
    Courses.find({ courseName: { $regex: coursePattern } })
        .then(course => {
            res.status(201).json(course);
        }).catch((e) => {
            console.log(e)
        })
});


router.get('/getUserCourse', reqLogin,async(req, res) => {
    await User.find({ _id: req.user._id })
    .populate("ratingsGiven" ," _id courseName channelName coursePic")
        .then((user) => {
            res.status(201).json(user);
        }).catch((e) => {
            console.log(e)
        })
});


router.get('/getCourse/:id', async(req, res) => {
    await Courses.find({ _id: req.params.id })
    .populate("comments","commentedText commentedBy")
    .then(course => {
        res.status(201).json(course);
    }).catch((e) => {
        console.log(e);
    })
});

router.get('/getCourseByCat/:category', async(req, res) => {
    await Courses.find({ category: req.params.category })
    .populate("comments","commentedText commentedBy")
    .then(course => {
        res.status(201).json({data:course});
    }).catch((e) => {
        console.log(e);
    })
});

router.put('/removeRating/:id', reqLogin,async(req, res) => {
    const data = await Courses.findByIdAndUpdate(req.user._id, {
        $pull: {
            ratedBy: req.user._id
        }
    });

    const removeFromUser = await User.findByIdAndUpdate(req.user._id, {
        $pull: {
            ratingGiven:req.params.id
        }
    })
    res.status(201).json(data, removeFromUser);
});

router.get('/reqCourses',reqLogin,reqAdmin,async(req,res)=>{
    try{
    const courses = await Courses.find({isAccepted:false});
    res.status(201).json({data:courses});
    }catch(e){
        res.status(401).json({error:"Error Occured"});
    }
});

router.put("/acceptCourse/:id",reqLogin,reqAdmin,async(req,res)=>{
    try{
        const accepted = await Courses.findByIdAndUpdate(req.params.id,{
            isAccepted:true
        });
        res.status(201).json({data:accepted}); 
    }catch(e){
        res.status(401).json({error:"Error Occured"});
    }
});

router.put("/rejectCourse/:id",reqLogin,reqAdmin,async(req,res)=>{
    try{
        await Courses.findByIdAndDelete(req.params.id);
        res.status(201).json({success:true,message:"Course Deleted"}); 
    }catch(e){
        res.status(401).json({error:"Error Occured"});
    }
});


router.delete("/deleteCourse/:id",reqLogin,reqAdmin,async(req,res)=>{
    try{
         await Courses.findByIdAndDelete(req.params.id);
        res.status(201).json({data:"Deleted"})
    }catch(e){
        res.status(401).json({error:"Error Occured"});
    }
});

router.get("/getCategory",async(req,res)=>{
    try{
        const category = await Category.find({});
        res.status(201).json({data:category});
    }catch(e){
        res.status(401).json({error:"Error Occured"});
    }
});


router.post("/addCategory",reqLogin,reqAdmin,async(req,res)=>{
    try{
        const {categoryName,categoryPic} = req.body;
        const category = Category({
            categoryName,
            categoryPic
        });

        await category.save().then((data)=>{
            res.status(201).json({data})
        })
    }catch(e){
        res.status(401).json({error:"Error Occured"});
    }
});

router.post("/addComment/:id",reqLogin,async(req,res)=>{
    try{
    const {commentedText}=req.body;
    if(!commentedText){
        return res.status(401).json({error:"Please Add Comment Text"});
    }else{
        const comment =  new Comment({
            commentedText,
            commentedBy:req.user.name,
            commentedCourse:req.params.id
        });
        await comment.save().then((d)=>{
            Courses.findByIdAndUpdate(req.params.id,{
                hasBeenCommented:true,
                $push:{
                    comments:d._id
                }
            }).then(()=>{
                res.status(201).json("Comment Added")
            })
        });
    }
}catch(e){
    res.status(401).json({error:"Error Occured"});
}
});

module.exports = router;